sdk/appcenter-identity/src/main/java/com/microsoft/appcenter/identity/storage/PreferenceTokenStorage.java

```diff
@@ -32,61 +69,161 @@
         mContext = context.getApplicationContext();
     }
 
-    /**
-     * Used for authentication requests, string field for auth token.
-     */
-    @VisibleForTesting
-    static final String PREFERENCE_KEY_AUTH_TOKEN = "AppCenter.auth_token";
+    @Override
+    public synchronized void saveToken(String token, String homeAccountId, Date expiresOn) {
+        List<TokenStoreEntity> history = getTokenHistory();
+        if (history == null) {
+            history = new ArrayList<TokenStoreEntity>() {{
 
-    /**
-     * Used for distinguishing users, string field for home account id.
-     */
-    @VisibleForTesting
-    static final String PREFERENCE_KEY_HOME_ACCOUNT_ID = "AppCenter.home_account_id";
+                /*
+                 * Adding a null entry is required during the first initialization to differentiate
+                 * anonymous usage before the moment and situation when we don't have a token
+                 * in history because of the size limit for example.
+                 */
+                add(new TokenStoreEntity(null, null, null));
+            }};
+        }
 
-    @Override
-    public void saveToken(String token, String homeAccountId) {
-        AuthTokenContext.getInstance().setAuthToken(token, homeAccountId);
-        String encryptedToken = CryptoUtils.getInstance(mContext).encrypt(token);
-        SharedPreferencesManager.putString(PREFERENCE_KEY_AUTH_TOKEN, encryptedToken);
-        SharedPreferencesManager.putString(PREFERENCE_KEY_HOME_ACCOUNT_ID, homeAccountId);
+        /* Do not add the same token twice in a row. */
+        String lastToken = history.size() > 0 ? history.get(history.size() - 1).getToken() : null;
+        if (!TextUtils.equals(lastToken, token)) {
+            history.add(new TokenStoreEntity(token, new Date(), expiresOn));
+        }
+
+        /* Limit history size. */
+        if (history.size() > TOKEN_HISTORY_LIMIT) {
+            history.remove(0);
+            AppCenterLog.debug(LOG_TAG, "Size of the token history is exceeded. The oldest token has been removed.");
+        }
+
+        /* Update history and current token. */
+        setTokenHistory(history);
+        if (token != null && homeAccountId != null) {
+            SharedPreferencesManager.putString(PREFERENCE_KEY_HOME_ACCOUNT_ID, homeAccountId);
+        } else {
+            SharedPreferencesManager.remove(PREFERENCE_KEY_HOME_ACCOUNT_ID);
+        }
     }
 
     @Override
-    public String getToken() {
-        String encryptedToken = SharedPreferencesManager.getString(PREFERENCE_KEY_AUTH_TOKEN, null);
-        if (encryptedToken == null || encryptedToken.length() == 0) {
-            return null;
+    public synchronized String getToken() {
+        List<TokenStoreEntity> history = getTokenHistory();
```

Comment:
I think there was a misunderstanding on my last comment about in memory.
We call this method often and this decrypts the full history every time.

What I said is that instead of having a persistent key just for latest token, we can cache in memory the latest value since its frequently used.

Is the reason we are afraid of memory dumps of some kind? If there is a security concern then might as well roll back that part to a single key for current, otherwise in memory.

The token will live in memory for offline calls anyway in http network state decorator, or all pending calls in general.

Reply:
Not anymore. See https://github.com/Microsoft/AppCenter-SDK-Android/pull/1006/files#diff-7ef491a2f4f5c31272e9f26cd400e0e8R61

Reply:
The part I missed is actually https://github.com/Microsoft/AppCenter-SDK-Android/pull/1006/files#diff-7ef491a2f4f5c31272e9f26cd400e0e8R165